Sofa/Component/Constraint/Lagrangian/Model/src/sofa/component/constraint/lagrangian/model/SlidingLagrangianConstraint.inl

```diff
@@ -57,109 +57,173 @@ void SlidingLagrangianConstraint<DataTypes>::init()
+            auto c1_it = c1->writeLine(cIndex);
+            c1_it.addCol(tm1, m_constraintDirections[i]);
 
-        auto c2_it = c2->writeLine(cIndex);
-        c2_it.addCol(tm2a, -m_dirAxe);
+            auto c2_it = c2->writeLine(cIndex);
+            c2_it.addCol(tm2a, -m_constraintDirections[i] * (1-m_bary));
+            c2_it.addCol(tm2b, -m_constraintDirections[i] * m_bary);
 
-        ++cIndex;
-    }
-    else if (r > ab)
-    {
-        m_thirdConstraint = r - ab;
+            ++cIndex;
+        }
 
-        auto c1_it = c1->writeLine(cIndex);
-        c1_it.setCol(tm1, -m_dirAxe);
+        auto c1_it_a = c1->writeLine(cIndex);
+        c1_it_a.setCol(tm1, m_constraintDirections[ m_constraintDirections.size() -2 ]);
+        auto c1_it_b = c1->writeLine(cIndex + 1);
+        c1_it_b.setCol(tm1, m_constraintDirections[ m_constraintDirections.size() -1 ]);
 
-        auto c2_it = c2->writeLine(cIndex);
-        c2_it.addCol(tm2b, m_dirAxe);
 
-        ++cIndex;
+        auto c2_it_a = c2->writeLine(cIndex);
+        c2_it_a.addCol(tm2a, -m_constraintDirections[ m_constraintDirections.size() -2 ]);
+        auto c2_it_b = c2->writeLine(cIndex + 1);
+        c2_it_b.addCol(tm2b, -m_constraintDirections[ m_constraintDirections.size() -1 ]);
+
+        cIndex += 2;
     }
 }
 
 
 template<class DataTypes>
-void SlidingLagrangianConstraint<DataTypes>::getConstraintViolation(const core::ConstraintParams *, linearalgebra::BaseVector *v, const DataVecCoord &, const DataVecCoord &
+void SlidingLagrangianConstraint<DataTypes>::getConstraintViolation(const core::ConstraintParams *, linearalgebra::BaseVector *v, const DataVecCoord & x1, const DataVecCoord & x2
         , const DataVecDeriv &, const DataVecDeriv &)
 {
-    const auto constraintIndex = this->d_constraintIndex.getValue();
+    const Coord_t<DataTypes> P = x1.getValue()[d_m1.getValue()];
+    const Coord_t<DataTypes> A = x2.getValue()[d_m2a.getValue()];
+    const Coord_t<DataTypes> B = x2.getValue()[d_m2b.getValue()];
 
-    v->set(constraintIndex, m_dist);
-    v->set(constraintIndex+1, 0.0);
+    const auto constraintIndex = this->d_constraintIndex.getValue();
 
-    if(m_thirdConstraint)
+    if(m_constraintDirections.size()==Deriv_t<DataTypes>::spatial_dimensions) // A and B are on the same point
+    {
+        //If A=B, we have bilateral constraint that fixed all dofs, they are thus aligned with the world system of coordinates and the violation is easily computed
+        typename DataTypes::CPos temp = DataTypes::getCPos(P)-DataTypes::getCPos(A);
+        for(unsigned i=0; i<m_constraintDirections.size() ; ++i)
+        {
+            v->set(constraintIndex + i, temp[i] );
+        }
+    }
+    else
     {
-        if(m_thirdConstraint>0)
-            v->set(constraintIndex+2, -m_thirdConstraint);
-        else
-            v->set(constraintIndex+2, m_thirdConstraint);
+        typename DataTypes::CPos newProj = DataTypes::getCPos(B)*m_bary + DataTypes::getCPos(A)*(1-m_bary);
```

Comment:
I am not sure the barycentric coordinate is the same in `buildConstraintMatrix` and in `getConstraintViolation`